worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala

```diff
@@ -1253,6 +1210,65 @@ class PushDataHandler(val workerSource: WorkerSource) extends BaseMessageHandler
     }
   }
 
+  private def writeLocalData(
+      fileWriters: Seq[FileWriter],
+      body: ByteBuf,
+      shuffleKey: String,
+      isPrimary: Boolean,
+      batchOffsets: Option[Array[Int]],
+      writePromise: Promise[Unit]): Unit = {
+    def writeData(fileWriter: FileWriter, body: ByteBuf, shuffleKey: String): Unit = {
+      try {
+        fileWriter.write(body)
+      } catch {
+        case e: Exception =>
+          if (e.isInstanceOf[AlreadyClosedException]) {
+            val (mapId, attemptId) = getMapAttempt(body)
+            val endedAttempt =
+              if (shuffleMapperAttempts.containsKey(shuffleKey)) {
+                shuffleMapperAttempts.get(shuffleKey).get(mapId)
+              } else -1
+            // TODO just info log for ended attempt
+            logWarning(s"Append data failed for task(shuffle $shuffleKey, map $mapId, attempt" +
+              s" $attemptId), caused by AlreadyClosedException, endedAttempt $endedAttempt, error message: ${e.getMessage}")
+          } else {
+            logError("Exception encountered when write.", e)
+          }
+          val cause =
+            if (isPrimary) {
+              StatusCode.PUSH_DATA_WRITE_FAIL_PRIMARY
+            } else {
+              StatusCode.PUSH_DATA_WRITE_FAIL_REPLICA
+            }
+          writePromise.failure(new CelebornIOException(cause))
+          fileWriter.decrementPendingWrites()
+      }
+    }
+    batchOffsets match {
+      case Some(batchOffsets) =>
+        var index = 0
+        var fileWriter: FileWriter = null
+        while (!writePromise.isCompleted && index < fileWriters.length) {
```

Comment:
I think we need to check `writePromise.isCompleted` in each loop, and if it is true, call `fileWriter.decrement`, as the original logic does:
```
        if (!alreadyClosed) {
          fileWriter.write(batchBody)
        } else {
          fileWriter.decrementPendingWrites()
        }
```

Reply:
updated to ```
        while (index < fileWriters.length) {
          if (!writePromise.isCompleted) {
            fileWriter = fileWriters(index)
            val offset = body.readerIndex() + batchOffsets(index)
            val length =
              if (index == fileWriters.length - 1) {
                body.readableBytes() - batchOffsets(index)
              } else {
                batchOffsets(index + 1) - batchOffsets(index)
              }
            val batchBody = body.slice(offset, length)
            writeData(fileWriter, batchBody, shuffleKey)
          } else {
            fileWriter.decrementPendingWrites()
          }
          index += 1
        }
```